hd-wallet/pluginv2.lua

```diff
@@ -0,0 +1,511 @@
+    local a = BigNum.from_bytes_be(Blob.from_hex(y))
+    local b = BigNum.from_bytes_be(Blob.from_hex("02"))
+    local c = BigNum.from_bytes_be(Blob.from_hex("00"))
+
+    if (a % b):to_bytes_be() == c:to_bytes_be() then
+        return "02"..x
+    else 
+        return "03"..x
+    end 
+end
+-- return public key from private key 
+function public_key_for_private_key(key_byte)
+    local point = get_point_coordinates_from_private_key(key_byte)
+    return compress_public_key(string.sub(point, 1, 64), string.sub(point, 65, 128))
+end
+-- parse input path
+function parse_path(child_path)
+    local path_table = split(child_path, "/")
+    return path_table
+end
+-- import chain-code as hmac key
+-- sign data as from hmac key
+function get_hmac(hmac_key, data)
+    local sobject = assert(Sobject.import { name = "hmac", obj_type = "HMAC", value = Blob.from_hex(hmac_key), transient = true })
+    local mac =  assert(sobject:mac { data = Blob.from_hex(data), alg = 'SHA512'}).digest
+    return mac:hex()
+end
+-- return ripmd160 digest
+function hash160(data)
+    local sha256_hash = assert(digest { data = Blob.from_hex(data), alg = 'SHA256' }).digest:hex()
+    local ripmd160_hash = assert(digest { data = Blob.from_hex(sha256_hash), alg = 'RIPEMD160' }).digest
+    return ripmd160_hash:hex()
+end
+-- add private keys
+function add_private_keys(k1, k2)
+    local a = BigNum.from_bytes_be(Blob.from_hex(k1))
+    local b = BigNum.from_bytes_be(Blob.from_hex(k2))
+    a:add(b)
+    a:mod(BigNum.from_bytes_be(Blob.from_hex(N)))   
+    hex_key = a:to_bytes_be():hex()
+    if (string.len( hex_key ) < 66) then
+        local offset = string.rep("0", 32-string.len( hex_key ))
+        hex_key = offset..hex_key
+    end
+    return hex_key
+end
+-- return scalar addition of point
+function add_public_keys(k1, k2)
+    local secP256K1 = EcGroup.from_name('SecP256K1')
+  	local pt_1 = secP256K1:point_from_binary(Blob.from_hex(k1))
+  	local pt_2 = secP256K1:point_from_binary(Blob.from_hex(k2))
+  	local p1 = secP256K1:point_from_components(pt_1:x(), pt_1:y())
+    local p2 = secP256K1:point_from_components(pt_2:x(), pt_2:y())
+    local p3 = p1 + p2
+    return compress_public_key(p3:x():to_bytes_be():hex(), p3:y():to_bytes_be():hex())
+end
+-- checksum: double sha256 of serialized key
+function get_check_sum(child_key)
+    local child_key_string = child_key.version.. child_key.depth..  child_key.fingerprint.. child_key.index.. child_key.chain_code.. child_key.key
+    local sha256_hash1 = assert(digest { data = Blob.from_hex(child_key_string), alg = 'SHA256' }).digest:hex()
+    local sha256_hash2 = assert(digest { data =  Blob.from_hex(sha256_hash1), alg = 'SHA256' }).digest:hex()
+    return sha256_hash2
+end
+-- derive new child key from parent key
+function derive_new_child(parent_key, child_idx)
+    local data = ""
+    -- if index is greater than equal to first hardend key
+    if tonumber(child_idx) >= FIRST_HARDENED_CHILD then
```

Comment:
Why did `FIRST_HARDENED_CHILD` changed from `0x80000000`? It only introduced more conversions such as this one.

Reply:
Looking at the code there is no justification I see, going back to hex